pkg/result/result_helper.go

```diff
@@ -1,20 +1,54 @@
 package result
 
 import (
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"time"
 
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
-// Copyright DataStax, Inc.
-// Please see the included license file for details.
-
+// This is just so that we can reference TerminalError in the Godoc of [Error]
+var _ error = reconcile.TerminalError(nil)
+
+// ReconcileResult represents the result of a step in the reconciliation process.
+//
+// We typically split the top-level Reconcile() method of a controller into multiple sub-functions. Each of these
+// functions uses ReconcileResult to communicate to its caller how the current iteration should proceed. There are 4
+// possible implementations: [Continue], [Done], [RequeueSoon], and [Error].
+//
+// The idiomatic way to handle a ReconcileResult in an intermediary sub-function is:
+//
+//	if recResult := callStep1(); recResult.Completed() {
+//	   // Global success, requeue or error: stop what we're doing and propagate up
```

Comment:
Issue: based on the logic you've described [here](https://github.com/k8ssandra/k8ssandra-operator/pull/1022#discussion_r1777359540) it sounded like a requeue (which is not a success) will also be `Completed()`?

Reply:
Yes, a requeue is `Completed()`:
```golang
func (c callBackSoon) Completed() bool {
	return true
}
```
But I don't see any issue here: if a step wants a requeue, we stop what we're doing and propagate up. The top-level Reconcile() will eventually return `recResult.Output()` (which is `ctrl.Result{Requeue: true, RequeueAfter: c.after}, nil`) to the controller runtime.

Reply:
Let me put my concern in another way. Right now, we have a struct implementation of this interface instantiated with `Done()` which is also `Completed()`. `Requeue()` and `Error()` are also `Completed()`. And yet, the English language meaning of the word `Completed()` is exactly the same as the meaning of the word `Done()`.

To me, having a getter function `Completed()` for a field across three different struct implementations ( `Done()`, `Requeue()` and `Error()` - some of which do not mean "Done" in the linguistic sense) is completely untenable. 

If you want to continue having something which means "this reconciliation run has progressed as far as possible, and needs to jump back to the top level of the reconciliation to be handled appropriately", we need a different word to `Completed()`. At that point, I think we can have a more productive conversation about what behaviours we are trying to model. But until we can express that intention in a function name, I would rather retain the explicit switch statements we have for `Done()`, `Requeue()` and `Error(), since this makes it explicit what behaviours we are handling.

Reply:
This is kind of why I think this needs a design doc too - because at the moment it isn't clear to me what behaviour or states we are trying to model, and we need to discuss and agree a target state rather than just say that we're "documenting" existing behaviour but then making quite impactful changes to the core of the reconciliation logic.

Reply:
> we need to discuss and agree a target state rather than just say that we're "documenting" existing behaviour

Yes, the goal of this PR is to have that discussion. I proposed an initial version, let's iterate on it with reviews until we reach a common understanding.

> this needs a design doc too

Are the go-docs not self-sufficient? Not that I mind having a separate document, but I wouldn't know what else to put in there.

> And yet, the English language meaning of the word `Completed()` is exactly the same as the meaning of the word `Done()`.

I've deliberately avoided going down the naming rabbithole so far... May I suggest that we keep that for a second step? I feel like it will get too messy if we rename as we go.
(but yes I agree with most of your comments)

Reply:
We already chatted about this on the call, but I think a design doc is a better place to facilitate discussion than a PR since it has better threaded comments etc. PRs for discussion will get messy, but it is good indeed to have some code to point at, so maybe both is best for this case.

Reply:
> Are the go-docs not self-sufficient? Not that I mind having a separate document, but I wouldn't know what else to put in there.

They are and should be the only place where API is documented. Everything else is out of the code and probably not maintained or reliable.

Reply:
This isn't about documentation, it is about the development process. We do deep discussions in design docs, not on PRs. If we choose to do it differently here, I'm concerned that we will end up in the same situation that we ended up in when we started making changes to the name overrides in labels and downstream resources, where the blast radius of the change was not adequately considered and customers suffered issues as a result. 

I strongly advocate for the creation of a design doc for this piece of work to avoid customer impacts.

Reply:
> And yet, the English language meaning of the word Completed() is exactly the same as the meaning of the word Done().

What is the English language definition difference of while and while do then? Yet, semantics for those two are different and in every programming language. One has while executed after first execution, while one has while executed before execution.

This isn't English.

Reply:
I agree with Miles, and let's not forget many of the people participating in these projects are not necessarily english native speakers. This API brings a level of confusion that is high enough to require having that conversation. There's overlap between `Completed()` (which should really be `IsCompleted()`), `Done()` and `Continue()` among other things which require to take a break and think of what this API is solving and how it should do it.
The fact that it's been used successfully (which is debatable) for the past few years isn't a valid reason to push back on the discussion.
I also think it's fair to have a separate doc, in the form of a markdown file, contributed as a PR so that the community can see it, to describe our view on the reconcile process and how `ReconcileResult` should be used.
"Design doc" might be an overstatement here and "doc" would probably be enough.

Reply:
> which should really be IsCompleted()

Ugly Java POJO, but this isn't a Java project. We don't need to indicate it's a boolean when a function has return value of boolean. Same as with "Get" when fetching an object.

> I also think it's fair to have a separate doc, in the form of a markdown file, contributed as a PR so that the community can see it

Why can't the users view what's defined in the godoc? It would be publicly available and generated already, like currently:

https://pkg.go.dev/github.com/k8ssandra/k8ssandra-operator@v1.20.3/pkg/reconciliation

What Olivier wrote here would appear there. We can add examples, descriptions etc. Adding markdown doc somewhere assumes users know to search for it (and hope it's not outdated) instead of the "standard" place.

Reply:
In an effort to move forward again, let's just use these in code comments as docs on how to use this interface.
The comments will be available in the IDE and in the godocs, without requiring to look at an external doc.
We can add more information here if we feel there's not enough for anyone to build a good understanding of how to use it.
We can extract this into a doc if that makes sense to anyone, but let's not make that mandatory.